src/backend/distributed/metadata/node_metadata.c

```diff
@@ -428,66 +428,141 @@ citus_disable_node(PG_FUNCTION_ARGS)
 {
 	text *nodeNameText = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
+	bool forceDisableNode = PG_GETARG_BOOL(2);
+
 	char *nodeName = text_to_cstring(nodeNameText);
 	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
-	bool isActive = false;
-	bool onlyConsiderActivePlacements = false;
-	MemoryContext savedContext = CurrentMemoryContext;
 
-	PG_TRY();
+	WorkerNode *firstWorkerNode = GetFirstPrimaryWorkerNode();
+	if (!forceDisableNode && firstWorkerNode &&
+		firstWorkerNode->nodeId == workerNode->nodeId)
 	{
-		if (NodeIsPrimary(workerNode))
-		{
-			/*
-			 * Delete reference table placements so they are not taken into account
-			 * for the check if there are placements after this.
-			 */
-			DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
+		/*
+		 * We sync metadata async and optionally in the background worker,
+		 * it would mean that some nodes might get the updates while other
+		 * not. And, if the node metadata that is changing is the first
+		 * worker node, the problem gets nasty. We serialize modifications
+		 * to replicated tables by acquiring locks on the first worker node.
+		 *
+		 * If some nodes get the metadata changes and some do not, they'd be
+		 * acquiring the locks on different nodes. Hence, having the
+		 * possibility of diverged shard placements for the same shard.
+		 *
+		 * To prevent that, we currently do not allow disabling the first
+		 * worker node.
+		 */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("disabling the first worker node in the "
+							   "metadata is not allowed"),
+						errhint("You can force disabling node, but this operation "
+								"might cause data loss: SELECT "
+								"citus_disable_node('%s', %d, force:=true);",
+								workerNode->workerName,
+								nodePort)));
+	}
 
-			if (NodeGroupHasShardPlacements(workerNode->groupId,
-											onlyConsiderActivePlacements))
-			{
-				ereport(NOTICE, (errmsg(
-									 "Node %s:%d has active shard placements. Some queries "
-									 "may fail after this operation. Use "
-									 "SELECT citus_activate_node('%s', %d) to activate this "
-									 "node back.",
-									 workerNode->workerName, nodePort,
-									 workerNode->workerName,
-									 nodePort)));
-			}
+	/*
+	 * First, locally mark the node as inactive. We'll later trigger background
+	 * worker to sync the metadata changes to the relevant nodes.
+	 */
+	bool isActive = false;
+	workerNode =
+		SetWorkerColumnLocalOnly(workerNode,
+								 Anum_pg_dist_node_isactive,
+								 BoolGetDatum(isActive));
+	if (NodeIsPrimary(workerNode))
+	{
+		/* there is no concept of invalid coordinator */
+		ErrorIfCoordinatorMetadataSetFalse(workerNode, BoolGetDatum(isActive),
+										   "isactive");
```

Comment:
Slightly odd to not just check this up front.

Reply:
I think it is mostly relevant to disabling secondary coordinator. But probably we should not allow disabled secondary of coordinator as well, so moving up.